sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java

```diff
@@ -1026,7 +1027,7 @@ public void start() throws JMSException {
         }
```

Comment:
may replace the direct assignment of producer with startProducer() thus make it clear that producer is opened in single code path.

Also, I see " isProducerNeedsToBeCreated" is removed in several places and connect() is only called in DoFn's setup. Can we get rid of this flag now?

Reply:
We added the flag in case of a failed connection it will create a new connection based on exceptionListener. Do you think it would be better to check with the producer if it's null ?

````java
if (producer == null) {
// open connection
connection.setExceptionListener(exception -> {
    connectionCounter.inc();
    // if there is an issue with the connection, we will close session, connection & producer so it can be recreated it
    close();
});
// create new producer
}
````

Reply:
Both is fine, if the flag is used then it should be consistent. For now if closeProducer() is called, producer is closed and set to null, but this flag is still true.